File: XUCore/debug/FCLog.m
// 
// FCLog.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-14 Charlie Monroe Software. All rights reserved.
// 


#import "FCLog.h"

#import <XUCore/XUCore-Swift.h>

NSString * __nonnull const FCLoggingStatusChangedNotification = @"FCLoggingStatusChangedNotification";

static BOOL _didCachePreferences = NO;
static BOOL _cachedPreferences = NO;
static BOOL _didRedirectToLogFile = NO;

static FILE *_logFile;

static NSTimeInterval _lastLogTimeInterval;

NSString *FCLogFilePath(){
	//The app identifier
	NSString *appIdentifier = [[NSBundle mainBundle] bundleIdentifier];
	if (appIdentifier == nil){
		//Probably a process
		appIdentifier = [[NSProcessInfo processInfo] processName];
	}
	
	//Make sure the logs directory exists
	NSString *logFolder = [[NSString stringWithFormat:@"~/Library/Application Support/%@/Logs/", appIdentifier] stringByExpandingTildeInPath];
	NSString *logFile = [logFolder stringByAppendingPathComponent:[NSString stringWithFormat:@"%@.log", appIdentifier]];
	[[NSFileManager defaultManager] createDirectoryAtPath:logFolder withIntermediateDirectories:YES attributes:nil error:nil];
	
	return logFile;
}

static inline void _FCCachePreferences(){
	if (!_didCachePreferences){
		_didCachePreferences = YES;
		_cachedPreferences = [[NSUserDefaults standardUserDefaults] boolForKey:@"FCLoggingEnabled"];
	}
}

static BOOL _FCRunningDevelopmentComputer(){
	#if !defined(DEBUG)
		#define DEBUG_REDEFINED
		#define DEBUG 0
	#endif
	return ([[[[NSProcessInfo processInfo] environment] objectForKey:@"PWD"] hasPrefix:@"/Users/oli/"]
		|| [[[[NSProcessInfo processInfo] environment] objectForKey:@"PWD"] hasPrefix:@"/Users/charlie/"]
		|| [[[[NSProcessInfo processInfo] environment] objectForKey:@"PWD"] hasPrefix:@"/Users/charliemonroe/"]
		|| TARGET_IPHONE_SIMULATOR
		|| (TARGET_OS_IPHONE && DEBUG));
	#if defined(DEBUG_REDEFINED)
		#undef DEBUG_REDEFINED
		#undef DEBUG
	#endif
}

static void _FCRedirectToLogFile() {
	// DO NOT LOG ANYTHING IN THIS FUNCTION,
	// AS YOU'D MAKE AN INFINITE LOOP!
	
	if (_FCRunningDevelopmentComputer()){
		return;
	}
	
	NSString *logFile = FCLogFilePath();
	
	// Try to create the log file
	if (![[NSFileManager defaultManager] fileExistsAtPath:logFile]){
		[(NSData*)[NSData data] writeToFile:logFile atomically:YES];
	}
	
	//Open the file
	_logFile = fopen([logFile fileSystemRepresentation], "a+");
	if (_logFile != NULL){
		//Making sure it exists
		int fileDesc = fileno(_logFile);
		dup2(fileDesc, STDOUT_FILENO);
		dup2(fileDesc, STDERR_FILENO);
		
		_didRedirectToLogFile = YES;
	}
}

static void _FCStartNewSession() {
	NSProcessInfo *processInfo = [NSProcessInfo processInfo];
	
	NSString *version = [[XUApplicationSetup sharedSetup] applicationVersionNumber];
	NSString *buildNumber = [[XUApplicationSetup sharedSetup] applicationBuildNumber];
	NSString *buildType = [[XUApplicationSetup sharedSetup] AppStoreBuild] ? @"AppStore" : @"Trial";
	
	NSLog(@"\n\n\n============== Starting a new %@ session (version %@[%@], %@) ==============", [processInfo processName], version, buildNumber, buildType);
}

static void _FCLogInitializer() __attribute__ ((constructor));
static void _FCLogInitializer() {
	if (_didCachePreferences){
		// No double-initialization
		return;
	}
	
	//Don't redirect the log on my computer
	if (_FCRunningDevelopmentComputer()){
		_didCachePreferences = YES;
		_cachedPreferences = YES;
		return;
	}
	
	_FCCachePreferences();
	if (_cachedPreferences){
		_FCRedirectToLogFile();
		_FCStartNewSession();
	}
}

void __FCLogSetShouldLog(BOOL log) {
	if (log && !_didRedirectToLogFile){
		_FCRedirectToLogFile();
		_FCStartNewSession();
	}
	
	BOOL didChange = log != _cachedPreferences;
	
	_cachedPreferences = log;
	_didCachePreferences = YES; //Already cached hence
	
	if (didChange) {
		[[NSNotificationCenter defaultCenter] postNotificationName:FCLoggingStatusChangedNotification object:nil];
	}
}

void FCForceSetDebugging(BOOL debug){
	__FCLogSetShouldLog(debug);
	
	[__XULogBridge setShouldLog:debug];
}

BOOL FCShouldLog(){
	return _cachedPreferences;
}

void FCForceLog(NSString *format, ...){
	BOOL originalCachedPrefs = _cachedPreferences;
	FCForceSetDebugging(YES);
	
	va_list argList;
	va_start(argList, format);
	NSLogv(format, argList);
	va_end(argList);

	_cachedPreferences = originalCachedPrefs;
}

void _FCLog(NSString *format, ...){
	if (!_didCachePreferences){
		_FCLogInitializer();
	}
	if (_cachedPreferences){
		va_list argList;
		va_start(argList, format);
		NSLogv(format, argList);
		va_end(argList);
		
		_lastLogTimeInterval = [NSDate timeIntervalSinceReferenceDate];
	}
}

void FCLogv(NSString *format, va_list args){
	if (_cachedPreferences){
		NSLogv(format, args);
	}
}

static inline NSString *__FCStacktraceStringFromSymbols(NSArray *symbols){
	NSMutableString *trace = [NSMutableString string];
	BOOL firstLine = YES;
	for (NSString *symbol in symbols) {
		if (firstLine){
			[trace appendString:symbol];
			firstLine = NO;
		}else{
			[trace appendFormat:@"\n%@", symbol];
		}
	}
	return trace;
}

NSString *FCStacktraceString(void){
	return __FCStacktraceStringFromSymbols([NSThread callStackSymbols]);
}

NSString *FCStacktraceStringFromException(NSException *exception){
	return __FCStacktraceStringFromSymbols([exception callStackSymbols]);
}

void FCLogStacktrace(NSString *comment){
	FCLog(@"%@: %@", comment, FCStacktraceString());
}

void FCClearLog(void){
	if (_logFile != NULL){
		fclose(_logFile);
		_logFile = NULL;
		
		[[NSFileManager defaultManager] removeItemAtPath:FCLogFilePath() error:NULL];
		
		_didRedirectToLogFile = NO;
		
		if (_cachedPreferences){
			_FCRedirectToLogFile();
			_FCStartNewSession();
		}
	}
}

NSDate *FCLastLogDate(void){
	if (_lastLogTimeInterval == 0.0){
		return nil;
	}
	return [NSDate dateWithTimeIntervalSinceReferenceDate:_lastLogTimeInterval];
}

File: XUCore/debug/FCLog.swift
//
//  FCLog.swift
//  DownieCore
//
//  Created by Charlie Monroe on 10/28/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Cocoa

let FCLoggingStatusChangedNotification = "FCLoggingStatusChangedNotification"

private var _cachedPreferences = false
private var _didCachePreferences = false
private var _didRedirectToLogFile = false
private var _lastLogTimeInterval: NSTimeInterval = NSDate.timeIntervalSinceReferenceDate()
private var _logFile: UnsafeMutablePointer<FILE>?


private func _FCCachePreferences() {
	if !_didCachePreferences {
		_didCachePreferences = true
		_cachedPreferences = NSUserDefaults.standardUserDefaults().boolForKey("FCLoggingEnabled")
	}
}
private func _FCRunningDevelopmentComputer() -> Bool {
	#if DEBUG
		return true
	#endif
	#if TARGET_OS_SIMULATOR
		return true
	#endif
	
	if let PWD = NSProcessInfo().environment["PWD"] {
		return PWD.hasPrefix("/Users/charliemonroe/")
	}
	
	return false
}
private func _FCRedirectToLogFile() {
	// DO NOT LOG ANYTHING IN THIS FUNCTION,
	// AS YOU'D MAKE AN INFINITE LOOP!
	
	if _FCRunningDevelopmentComputer() {
		return
	}
	
	let logFile = FCLogFilePath()
	
	// Try to create the log file
	if !NSFileManager.defaultManager().fileExistsAtPath(logFile) {
		NSData().writeToFile(logFile, atomically: true)
	}
	
	_logFile = fopen((logFile as NSString).fileSystemRepresentation, ("a+" as NSString).UTF8String)
	if _logFile != nil {
		let fileDesc = fileno(_logFile!)
		dup2(fileDesc, STDOUT_FILENO)
		dup2(fileDesc, STDERR_FILENO)
		_didRedirectToLogFile = true
	}
}
private func _FCStartNewSession() {
	let processInfo = NSProcessInfo()
	let mainBundle = NSBundle.mainBundle()
	
	let version = mainBundle.objectForInfoDictionaryKey("CFBundleShortVersionString") as? String ?? "1.0"
	let buildNumber = mainBundle.objectForInfoDictionaryKey("CFBundleVersion") as? String ?? "0"
	
	print("\n\n\n============== Starting a new \(processInfo.processName) session (version \(version)[\(buildNumber)]) ==============")
}
private func _FCLogInitializer() {
	if _didCachePreferences {
		// No double-initialization
		return
	}
	
	//Don't redirect the log on my computer
	if _FCRunningDevelopmentComputer(){
		_didCachePreferences = true
		_cachedPreferences = true
		return
	}
	
	_FCCachePreferences()

	if _cachedPreferences {
		_FCRedirectToLogFile()
		_FCStartNewSession()
	}
}


func FCLogFilePath() -> String {
	let appIdentifier = NSBundle.mainBundle().bundleIdentifier ?? NSProcessInfo().processName
	
	let logFolder = ("~/Library/Application Support/\(appIdentifier)/Logs/" as NSString).stringByExpandingTildeInPath
	let logFile = logFolder + "/" + "\(appIdentifier).log"
	_ = try? NSFileManager.defaultManager().createDirectoryAtPath(logFolder, withIntermediateDirectories: true, attributes: nil)
	return logFile
}

func FCForceLog(@autoclosure string: () -> String, method: String = __FUNCTION__, file: String = __FILE__, line: Int = __LINE__) {
	let originalPreferences = _cachedPreferences
	_cachedPreferences = true
	FCLog(string, method: method, file: file, line: line)
	_cachedPreferences = originalPreferences
}

func FCLog(@autoclosure string: () -> String, method: String = __FUNCTION__, file: String = __FILE__, line: Int = __LINE__) {
	if !_didCachePreferences {
		_FCLogInitializer()
	}
	
	if _cachedPreferences {
		print("\((file as NSString).lastPathComponent):\(line).\(method): \(string())")
		
		_lastLogTimeInterval = NSDate.timeIntervalSinceReferenceDate()
	}
}

func FCForceSetDebugging(debug: Bool) {
	if debug && !_didRedirectToLogFile {
		_FCRedirectToLogFile()
		_FCStartNewSession()
	}
	
	let didChange = debug != _cachedPreferences;
	
	_cachedPreferences = debug
	_didCachePreferences = true //Already cached hence
	
	if (didChange) {
		NSNotificationCenter.defaultCenter().postNotificationName(FCLoggingStatusChangedNotification, object: nil)
	}
}
func FCShouldLog() -> Bool {
	return _cachedPreferences
}

func FCClearLog() {
	if (_logFile != nil){
		fclose(_logFile!);
		_logFile = nil;
		
		_ = try? NSFileManager.defaultManager().removeItemAtPath(FCLogFilePath())
		
		_didRedirectToLogFile = false
		
		if _cachedPreferences {
			_FCRedirectToLogFile()
			_FCStartNewSession()
		}
	}
}
